utxo/src/undo.rs

```diff
@@ -46,26 +62,47 @@ impl TxUndo {
     }
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
+#[derive(Default, Debug, Clone, Eq, PartialEq, Encode, Decode)]
 pub struct BlockUndo {
-    // determines at what height this undo file belongs to.
-    height: BlockHeight,
-    undos: Vec<TxUndo>,
+    reward_undo: Option<BlockRewardUndo>,
+    tx_undos: Vec<TxUndo>,
 }
 
 impl BlockUndo {
-    pub fn new(tx_undos: Vec<TxUndo>, height: BlockHeight) -> Self {
+    pub fn new(reward_undo: Option<BlockRewardUndo>, tx_undos: Vec<TxUndo>) -> Self {
         Self {
-            height,
-            undos: tx_undos,
+            tx_undos,
+            reward_undo,
         }
     }
+
     pub fn tx_undos(&self) -> &[TxUndo] {
-        &self.undos
+        &self.tx_undos
+    }
+
+    pub fn push_tx_undo(&mut self, tx_undo: TxUndo) {
+        self.tx_undos.push(tx_undo);
     }
 
-    pub fn height(&self) -> BlockHeight {
-        self.height
+    pub fn take_tx_undo(&mut self, tx_num: usize) -> Option<TxUndo> {
+        if tx_num < self.tx_undos.len() {
+            Some(self.tx_undos.remove(tx_num))
```

Comment:
Ignoring the performance issues here, this can easily become a fire hazard. If you remove a transaction at an index other than the end, it basically means you've modified all the indices that come after it. Now we can look at this in two different ways. We either want to prevent this at the API level and return an error with `Result<Option<TxUndo>, Error>`, or consider this an invariant and should never happen, in which case we panic (aka, assert). For now, either way is fine for me.